spec/controllers/hyrax/stats_controller_spec.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe Hyrax::StatsController, type: :controller do
+  routes { HykuAddons::Engine.routes }
+
+  let(:user) { create(:admin) }
+  let(:account) { create(:account) }
+  let(:site) { Site.new(account: account) }
+  let(:gds_fixture) { File.read(HykuAddons::Engine.root.join("spec", "fixtures", "csv", "gds_chart_fixture.csv")) }
+  let(:enabled) { true }
+
+  before do
+    allow(Site).to receive(:instance).and_return(site)
+  end
+
+  describe "GET #reports" do
+    render_views
+
+    before do
+      allow(Flipflop).to receive(:enabled?).and_call_original
+      allow(Flipflop).to receive(:enabled?).with(:gds_reports).and_return(enabled)
+      allow(controller.current_ability).to receive(:can?).with(any_args).and_return(true)
+      allow(account).to receive(:gds_reports).and_return(gds_fixture)
+
+      sign_in(user)
+    end
+
+    context "when the user is an admin" do
+      it "renders a partial" do
+        get :reports
+
+        expect(response).to have_http_status(:ok)
+        expect(response).to render_template(:reports)
+      end
+
+      context "when there is data" do
+        before do
+          get :reports
+        end
+
+        it "displays the content" do
+          expect(response.body).to include(I18n.t("stats.reports.title"))
+          expect(response.body).to include(I18n.t("stats.reports.intro"))
+        end
+
+        it "displays the select" do
+          expect(response.body).to have_selector("#dashaboard_gds_reports")
+        end
+
+        it "displays the correct report titles in the select" do
+          reports = response.body.scan(/<option.*>(.*)<\/option>/).flatten
+          titles = gds_fixture.lines.map { |line| line.split(",").map(&:strip).delete_at(0) }
+
+          expect(reports).to match_array(titles)
+        end
+      end
+
+      context "when no data has been entered" do
+        let(:gds_fixture) { nil }
+
+        it "does not display the content" do
+          get :reports
+
+          expect(response.body).to include(I18n.t("stats.reports.title"))
+          expect(response.body).to include(I18n.t("stats.reports.no_reports"))
+        end
+      end
+
+      context "when the user is a non-admin" do
+        let(:enabled) { false }
+
+        it "renders a partial" do
+          expect { get :reports }.to raise_error(ActionController::RoutingError)
```

Comment:
Does this routing error get rescued so the user is redirected?  Do we also need to test that a link a non-admin user can't access isn't displayed in a menu or such?

Reply:
I would suggest its fine for a user to see a 404 for a page that they are unable to access. As these are controller specs they're not really designed for testing out the content of a page, so for a non-admin user I would have to perform a GET to another controller action in order to see if the link they could click is hidden from them